### ISSUE TITLE
Token renewal?

### ISSUE THREAD

**Issue by daitj** [laur89/revolut-py#1]
I am using your branch `feature/change-to-web-api` to check my revolut balance periodically, it seems like when the token expires, it again asks for code from the email(channel)? Is it possible to renew the token without any interaction?

**Comment by laur89** [comment]
Yes and no. This is the reason constructor accepts `provider_2fa` param. Easiest/most reasonable would be to use EMAIL for the `channel`, and pass a function as provider_2fa that resolves the code from your email. We still need to resolve the code, but at least it can be automated.

Theoretically should be doable for SMS channel as well, but then we'd have to route phone SMS somewhere where our logic is able to read them. Unless you're running your logic from phone itself :)

**Comment by daitj** [comment]
Thanks for the info, I wrote a quick script to check a particular folder from my gmail account using "App password". 

Leaving this here for anyone coming by here:

```py
# -*- coding: utf-8 -*-

import imaplib
import email
import time
from email.header import decode_header
from email.utils import parsedate_tz, mktime_tz, formatdate
import os
import re

# total time this function will run =  
#   2 * (retry_count * sleep_before_retry_seconds)

# time stuff
retry_count = 6
sleep_before_retry_seconds = 20

time_diff_check_seconds = retry_count * sleep_before_retry_seconds

# account credentials
username = "EMAIL@EXAMPLE.COM"
password = "PASSWORD"
imap_server = "imap.gmail.com"
N = 1

regex = r"email\/confirm\?token=([0-9]{3}-[0-9]{3})&phone"

def get_code_from_email(phone=None):
    loop_retry = 0
    waiting_for_second_code = False
    code = 0
    while True:
        time.sleep(sleep_before_retry_seconds)
        if waiting_for_second_code:
            # revolut trying to send second code to throw off the automated bots
            # check for second message after time_diff_check_seconds + the retry seconds above
            # this should be enough to fetch the second code if there is any
            print("waiting for second code (if there is)")
            time.sleep(time_diff_check_seconds)
        imap = imaplib.IMAP4_SSL(imap_server)
        imap.login(username, password)
        status, messages = imap.select("[Imap]/Revolut")
        messages = int(messages[0])
        print("messages count {}".format(messages))
        for i in range(messages, messages-N, -1):
            # fetch the email message by ID
            res, msg = imap.fetch(str(i), "(RFC822)")
            for response in msg:
                if isinstance(response, tuple):
                    # parse a bytes email into a message object
                    msg = email.message_from_bytes(response[1])
                    received_date, encoding = decode_header(msg.get("Date"))[0]
                    if isinstance(received_date, bytes):
                        received_date = received_date.decode(encoding)
                    tt = parsedate_tz(received_date)
                    timestamp = mktime_tz(tt)
                    time_diff = int(time.time())-timestamp
                    body = msg.get_payload(decode=True).decode()
                    if time_diff <= time_diff_check_seconds:
                        matches = re.search(regex, body)
                        if matches:
                            code = matches.group(1)
                            if waiting_for_second_code:
                                print("second code found and returned", code)
                                return code
                            else:
                                print("first code found", code)
                                waiting_for_second_code = True
                                continue
                    else:
                        if waiting_for_second_code:
                            # we have already got code in first email, second email was not there
                            # return the code we got in first email
                            print("first code returned")
                            return code
                        else:
                            print("old messages only, in try", loop_retry+1)

        # close the connection and logout
        imap.close()
        imap.logout()
        loop_retry+=1
        if loop_retry == retry_count:
            print("code not found, even after trying")
            return None

if __name__ == '__main__':
	get_code_from_email()
```

**Comment by laur89** [comment]
I know it's just a quick POC/example, but I'd add the `time.sleep()` to the start of the loop, not end. Best test and see how it handles in real life, but most likely mail still takes some time to be delivered.

**Comment by daitj** [comment]
Thanks for the tip. Like you said, the mail timing was not consistent. I had to change to bit higher values. 

Currently I cannot do a new login request. It fails with HTTP 429 error. I will test it later and post results here.

**Comment by daitj** [comment]
After some tests, I found out that, revolut sends two emails with code once in a while. One with invalid code and after a minute another one with a valid code. I didn't check if this was actually caused by multiple login requests from your code or just that the revolut is trying to honeypot the bots.

I updated my email fetching code to workaround this issue.

In theory this bot could run every minute for 30 mins when same login session is valid and then 5 mins cooldown to get new code and login session. 

For me I just use it to check my balance and notify me when it is running low so I can topup from my bank.
